File: Room/app/src/main/java/com/example/room/database/UserDAO.java
package com.example.room.database;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.example.room.model.Address;
import com.example.room.model.User;
import com.example.room.model.UserAddress;

import java.util.List;

@Dao
public interface UserDAO {

    @Insert
    void insertUser(User user);

    @Insert
    void insertAddress(Address address);

    @Update
    void updateUser(User user);

    @Update
    void updateAddress(Address address);

    @Transaction
    @Query("SELECT * FROM user")
    List<UserAddress> getAllUser();

    @Transaction
    @Query("SELECT * FROM  user WHERE user_id = :userId")
    UserAddress getUser(int userId);

    @Delete
    void deleteUser(User user);

    @Query("DELETE FROM user")
    void deleteAllUser();

    @Transaction
    @Query("SELECT * FROM user WHERE userName LIKE '%' || :name || '%'")
    List<UserAddress> searchUser(String name);
}

File: DataStorage/Realm/app/src/main/java/com/example/realm/model/UserDao.kt
package com.example.realm.model

import io.realm.Realm
import io.realm.RealmResults
import java.util.*

class UserDao {
    private val realm = Realm.getDefaultInstance()

    fun insert(user : User){
        realm.executeTransaction( Realm.Transaction {
            it.insertOrUpdate(user)
        })
    }

    fun getAll() : List<User>{
        val realmResults = realm
            .where(User::class.java)
            .findAll()
        var results : List<User>  = ArrayList<User>(realmResults)
        return results
    }

    fun delete(id : Int){
        realm.executeTransaction( Realm.Transaction {
            val realmResults : RealmResults<User>? = realm
                .where(User::class.java)
                .equalTo("id",id)
                .findAll()
            realmResults?.deleteAllFromRealm()
        })
    }

    fun deleteAll(){
        realm.executeTransaction( Realm.Transaction {
            val realmResults : RealmResults<User>? = realm
                .where(User::class.java)
                .findAll()
            realmResults?.deleteAllFromRealm()
        })
    }

    fun close(){
        realm?.close()
    }
}